File: studio/AuraStudio/Sources/AuraStudio/Services/LibraryEnricher.swift
import Foundation

/// Adivina artista/titulo a partir de un nombre de archivo cuando no
/// hay tags previas -- el patron mas comun con diferencia es
/// "Artista - Titulo.ext". Si no matchea ese patron, se usa el nombre
/// completo (sin extension) como titulo y se busca solo por eso.
enum FilenameGuesser {
    static func guess(from url: URL) -> (artist: String?, title: String?) {
        let base = url.deletingPathExtension().lastPathComponent
        let parts = base.components(separatedBy: " - ")
        if parts.count >= 2, !looksLikeTrackNumberPrefix(parts[0]) {
            return (parts[0].trimmingCharacters(in: .whitespaces), parts.dropFirst().joined(separator: " - ").trimmingCharacters(in: .whitespaces))
        }
        return (nil, base.trimmingCharacters(in: .whitespaces))
    }

    /// PLAN-sync-media-hardening.md PARTE 1A: el primer segmento antes de
    /// " - " a veces es en realidad el NUMERO DE PISTA pegado al nombre
    /// del archivo ("1 - Titulo.m4a", "01 Lil Dub Chefin - Version...
    /// m4a"), no el artista -- visto en produccion: decenas de canciones
    /// sin tags de artista terminaron en carpetas del iPod literalmente
    /// llamadas "1".."20" (una por numero de pista, mezclando canciones
    /// de artistas distintos), en vez de "Desconocido". Un segmento que
    /// arranca con 1-3 digitos seguidos de un espacio (o que es
    /// puramente numerico) no es un nombre de artista plausible --
    /// heuristica imperfecta a proposito (un artista real como "21
    /// Savage" cae a "Desconocido" en vez de a su nombre), pero muchisimo
    /// mejor que agrupar decenas de artistas bajo una carpeta "1".
    static func looksLikeTrackNumberPrefix(_ segment: String) -> Bool {
        let trimmed = segment.trimmingCharacters(in: .whitespaces)
        guard let firstNonDigitIndex = trimmed.firstIndex(where: { !$0.isNumber }) else {
            return !trimmed.isEmpty
        }
        let digitCount = trimmed.distance(from: trimmed.startIndex, to: firstNonDigitIndex)
        guard digitCount >= 1 && digitCount <= 3 else { return false }
        return trimmed[firstNonDigitIndex] == " "
    }
}

/// Orquesta el flujo "arrastrar y listo" para un LibraryItem de musica:
/// parte de lo que ya tenga el archivo (tags ID3 existentes, si es
/// MP3), completa lo que falte via MusicBrainz/Cover Art Archive/LRCLIB,
/// y devuelve metadata lista para escribir. No copia nada al iPod
/// (eso es tarea de LibrarySync, despues de que el usuario revise/
/// confirme si quiere).
struct LibraryEnricher {
    let musicBrainz: MusicBrainzClient
    let coverArt: CoverArtArchiveClient
    let lrclib: LRCLIBClient
    let fanartTV: FanartTVClient
    let deezer: DeezerClient

    init(musicBrainz: MusicBrainzClient = MusicBrainzClient(),
         coverArt: CoverArtArchiveClient = CoverArtArchiveClient(),
         lrclib: LRCLIBClient = LRCLIBClient(),
         fanartTV: FanartTVClient = FanartTVClient(),
         deezer: DeezerClient = DeezerClient()) {
        self.musicBrainz = musicBrainz
        self.coverArt = coverArt
        self.lrclib = lrclib
        self.fanartTV = fanartTV
        self.deezer = deezer
    }

    /// D-203: prueba los proveedores de portada en el orden que eligio
    /// el usuario (Ajustes > Servicios) y devuelve la primera imagen que
    /// aparezca. `fanartTV`/`deezer` ya devuelven nil solos cuando no
    /// corresponde intentarlos (sin key, o busqueda sin resultado) --
    /// aca solo hace falta saltear Deezer si el usuario lo apago, y
    /// tragar cualquier error de red con `try?` para que un proveedor
    /// que falla no tumbe a los demas.
    private func resolveCoverArt(releaseID: String, releaseGroupID: String?,
                                  title: String?, artist: String?,
                                  order: [CoverArtProvider], deezerEnabled: Bool) async -> Data? {
        for provider in order {
            switch provider {
            case .coverArtArchive:
                if let data = try? await coverArt.fetchFrontCover(releaseID: releaseID) { return data }
            case .fanartTV:
                guard let releaseGroupID else { continue }
                if let data = try? await fanartTV.fetchAlbumCover(releaseGroupID: releaseGroupID) { return data }
            case .deezer:
                guard deezerEnabled, let title, let artist else { continue }
                if let data = try? await deezer.fetchAlbumCover(title: title, artist: artist) { return data }
            }
        }
        return nil
    }

    /// Umbral minimo de `score` de MusicBrainz (0-100) para aceptar una
    /// `Recording` como fuente de album/año -- `MusicBrainzClient.
    /// searchRecording` siempre devuelve el resultado de mayor score
    /// aunque sea bajo (o `nil`), y `enrich()`/`reenrich()` lo usaban
    /// sin piso: dos canciones del mismo album real podian terminar con
    /// albumes distintos si MusicBrainz devolvia un release equivocado
    /// con score bajo. Sin tags locales, es mejor dejar "Sin album"
    /// (revisable) que inventar uno. Un `score` ausente (`nil`) se trata
    /// como 0 -- por debajo del umbral, se rechaza.
    static let minimumMusicBrainzScore = 70

    /// `online` y `lyrics` salen de las preferencias del usuario
    /// (AppPreferences): con `online` en false no se toca la red y solo
    /// se usa lo que ya trae el archivo mas lo que se adivine del nombre.
    func enrich(item: LibraryItem, online: Bool = true, lyrics: Bool = true,
                coverArtOrder: [CoverArtProvider] = CoverArtProvider.defaultOrder,
                deezerEnabled: Bool = true) async -> TrackMetadata {
        var metadata = await LocalTagReader.readTag(from: item.sourceURL)

        let guess = FilenameGuesser.guess(from: item.sourceURL)
        metadata.title = metadata.title ?? guess.title
        metadata.artist = metadata.artist ?? guess.artist

        guard online else { return normalizingCover(metadata) }

        guard let recording = try? await musicBrainz.searchRecording(title: metadata.title, artist: metadata.artist),
              (recording.score ?? 0) >= Self.minimumMusicBrainzScore else {
            return normalizingCover(metadata)
        }

        metadata.title = metadata.title ?? recording.title
        metadata.artist = metadata.artist ?? recording.artistCredit?.first?.name
        metadata.musicBrainzRecordingID = recording.id

        if let release = recording.releases?.first {
            metadata.album = metadata.album ?? release.title
            metadata.year = metadata.year ?? release.date.map { String($0.prefix(4)) }
            metadata.musicBrainzReleaseID = release.id

            if metadata.coverArtData == nil {
                metadata.coverArtData = await resolveCoverArt(
                    releaseID: release.id, releaseGroupID: release.releaseGroup?.id,
                    title: metadata.title, artist: metadata.artist,
                    order: coverArtOrder, deezerEnabled: deezerEnabled)
            }
        }

        if lyrics, let title = metadata.title, let artist = metadata.artist {
            metadata.syncedLyrics = try? await lrclib.fetchSyncedLyrics(title: title, artist: artist, album: metadata.album)
        }

        return normalizingCover(metadata)
    }

    /// ST-141: toda carátula que entra a la biblioteca queda cuadrada,
    /// venga de la etiqueta del archivo (`LocalTagReader`, incluido el
    /// `cover.jpg` de la carpeta) o de la red. Acá está el único punto
    /// por el que pasan las dos.
    private func normalizingCover(_ metadata: TrackMetadata) -> TrackMetadata {
        guard let cover = metadata.coverArtData else { return metadata }
        var normalized = metadata
        normalized.coverArtData = CoverArtNormalizer.normalized(cover)
        return normalized
    }

    /// D-198: "Buscar información"/"Buscar letra" del menú contextual de
    /// la biblioteca -- a diferencia de `enrich()` (que parte de las
    /// tags CRUDAS del archivo original), esto parte de la metadata YA
    /// resuelta del item (`currentMetadata`), para no pisar una
    /// corrección manual que el usuario ya hizo en la pantalla de
    /// revisión. Solo llena huecos (nunca reemplaza un campo que ya
    /// tiene valor) -- mismo criterio que `enrich()`.
    ///
    /// D-203: a diferencia de la version anterior, esto YA NO traga los
    /// errores de red en silencio -- devuelve un `EnrichmentOutcome`
    /// junto con la metadata para que quien llama (`LibraryViewModel`)
    /// le pueda mostrar al usuario "no se encontro nada" vs "fallo la
    /// conexion" en vez de que ambos casos se vean identicos (la causa
    /// mas probable de que el dueño reportara que esto "no sirve para
    /// nada": un error de red silencioso es indistinguible de que de
    /// verdad no habia informacion que agregar).
    func reenrich(item: LibraryItem, currentMetadata: TrackMetadata,
                  fetchAlbumInfo: Bool, fetchLyrics: Bool,
                  coverArtOrder: [CoverArtProvider] = CoverArtProvider.defaultOrder,
                  deezerEnabled: Bool = true) async -> (metadata: TrackMetadata, outcome: EnrichmentOutcome) {
        var metadata = currentMetadata
        var outcome = EnrichmentOutcome()
        let guess = FilenameGuesser.guess(from: item.sourceURL)
        let seedTitle = metadata.title ?? guess.title
        let seedArtist = metadata.artist ?? guess.artist

        if fetchAlbumInfo {
            do {
                if let recording = try await musicBrainz.searchRecording(title: seedTitle, artist: seedArtist),
                   (recording.score ?? 0) >= Self.minimumMusicBrainzScore {
                    outcome.albumInfoFound = true
                    metadata.title = metadata.title ?? recording.title
                    metadata.artist = metadata.artist ?? recording.artistCredit?.first?.name
                    metadata.musicBrainzRecordingID = metadata.musicBrainzRecordingID ?? recording.id

                    if let release = recording.releases?.first {
                        metadata.album = metadata.album ?? release.title
                        metadata.year = metadata.year ?? release.date.map { String($0.prefix(4)) }
                        metadata.musicBrainzReleaseID = metadata.musicBrainzReleaseID ?? release.id

                        if metadata.coverArtData == nil {
                            metadata.coverArtData = await resolveCoverArt(
                                releaseID: release.id, releaseGroupID: release.releaseGroup?.id,
                                title: metadata.title, artist: metadata.artist,
                                order: coverArtOrder, deezerEnabled: deezerEnabled)
                        }
                    }
                }
            } catch {
                outcome.networkErrorMessage = error.localizedDescription
            }
        }

        if fetchLyrics, let title = metadata.title, let artist = metadata.artist {
            do {
                metadata.syncedLyrics = try await lrclib.fetchSyncedLyrics(title: title, artist: artist, album: metadata.album)
                outcome.lyricsFound = metadata.syncedLyrics != nil
            } catch {
                outcome.networkErrorMessage = outcome.networkErrorMessage ?? error.localizedDescription
            }
        }

        return (normalizingCover(metadata), outcome)
    }
}

/// Resultado de `reenrich()` para un item -- que tanto encontro, y si
/// algo salio mal a nivel de red (vs. simplemente "no habia nada que
/// encontrar", que no es un error).
struct EnrichmentOutcome: Equatable {
    var albumInfoFound = false
    var lyricsFound = false
    var networkErrorMessage: String?
}

File: studio/windows/AuraStudio.Core/Library/LibraryEnricher.cs
using AuraStudio.Core.Networking;

namespace AuraStudio.Core.Library;

/// <summary>
/// De dónde puede salir la carátula de un álbum. El usuario elige el orden en
/// Ajustes → Servicios (D-203) y se prueba en ese orden hasta que aparezca una.
/// </summary>
public enum CoverArtProvider { CoverArtArchive, FanartTV, Deezer }

public static class CoverArtProviderInfo
{
    public static readonly IReadOnlyList<CoverArtProvider> DefaultOrder =
        [CoverArtProvider.CoverArtArchive, CoverArtProvider.FanartTV, CoverArtProvider.Deezer];

    public static string DisplayName(this CoverArtProvider provider) => provider switch
    {
        CoverArtProvider.CoverArtArchive => "Cover Art Archive",
        CoverArtProvider.FanartTV => "fanart.tv",
        _ => "Deezer"
    };
}

/// <summary>
/// Qué tanto encontró un enriquecimiento, y si algo falló <b>a nivel de red</b>.
///
/// <para>D-203: no es lo mismo "no había nada que encontrar" que "se cayó la
/// conexión", y tragarse el error hacía que los dos casos se vieran idénticos —
/// la causa más probable de que el dueño reportara que esto "no sirve para
/// nada".</para>
/// </summary>
public sealed record EnrichmentOutcome
{
    public bool AlbumInfoFound { get; set; }
    public bool LyricsFound { get; set; }
    public string? NetworkErrorMessage { get; set; }
}

/// <summary>
/// Adivina artista y título del nombre del archivo cuando no hay etiquetas. El
/// patrón más común con diferencia es "Artista - Título.ext"; si no coincide, el
/// nombre completo se usa como título. Port de <c>FilenameGuesser</c>.
/// </summary>
public static class FilenameGuesser
{
    public readonly record struct Guess(string? Artist, string? Title);

    public static Guess For(string path)
    {
        string baseName = Path.GetFileNameWithoutExtension(path);
        string[] parts = baseName.Split(" - ");

        if (parts.Length >= 2 && !LooksLikeTrackNumberPrefix(parts[0]))
            return new Guess(parts[0].Trim(), string.Join(" - ", parts[1..]).Trim());

        return new Guess(null, baseName.Trim());
    }

    /// <summary>
    /// Si el primer segmento antes de " - " es en realidad el <b>número de
    /// pista</b> pegado al nombre ("1 - Título.m4a"), no el artista.
    ///
    /// <para>Se vio en producción: decenas de canciones sin etiqueta de artista
    /// terminaron en carpetas del iPod llamadas literalmente "1".."20" —una por
    /// número de pista, mezclando artistas distintos— en vez de en
    /// "Desconocido".</para>
    ///
    /// <para>La heurística es imperfecta <b>a propósito</b>: un artista real
    /// como "21 Savage" cae en "Desconocido" en vez de en su nombre, que es
    /// muchísimo mejor que agrupar decenas de artistas bajo una carpeta "1".</para>
    /// </summary>
    public static bool LooksLikeTrackNumberPrefix(string segment)
    {
        string trimmed = segment.Trim();

        int firstNonDigit = -1;
        for (int i = 0; i < trimmed.Length; i++)
            if (!char.IsDigit(trimmed[i])) { firstNonDigit = i; break; }

        // Puramente numérico: "01", "7". No es un nombre de artista plausible.
        if (firstNonDigit < 0) return trimmed.Length > 0;

        if (firstNonDigit is < 1 or > 3) return false;
        return trimmed[firstNonDigit] == ' ';
    }
}

/// <summary>
/// Orquesta el "arrastrar y listo" de una canción: parte de lo que ya trae el
/// archivo, completa lo que falte con MusicBrainz / Cover Art Archive / LRCLIB,
/// y devuelve metadata lista para escribir. Port de <c>LibraryEnricher.swift</c>.
///
/// <para><b>Solo llena huecos.</b> Nunca reemplaza un campo que ya tiene valor:
/// eso es lo que protege una corrección que el usuario hizo a mano.</para>
///
/// <para>No copia nada al iPod — eso es de la sincronización, después de que el
/// usuario revise.</para>
/// </summary>
public sealed class LibraryEnricher(
    MusicBrainzClient? musicBrainz = null,
    CoverArtArchiveClient? coverArt = null,
    LRCLIBClient? lrclib = null,
    FanartTVClient? fanartTV = null,
    DeezerClient? deezer = null,
    Func<string, Task<TrackMetadata>>? readTag = null)
{
    private readonly MusicBrainzClient _musicBrainz = musicBrainz ?? new MusicBrainzClient();
    private readonly CoverArtArchiveClient _coverArt = coverArt ?? new CoverArtArchiveClient();
    private readonly LRCLIBClient _lrclib = lrclib ?? new LRCLIBClient();
    private readonly FanartTVClient _fanartTV = fanartTV ?? new FanartTVClient();
    private readonly DeezerClient _deezer = deezer ?? new DeezerClient();

    /// <summary>La lectura local de etiquetas, inyectable para poder probar sin archivos.</summary>
    private readonly Func<string, Task<TrackMetadata>> _readTag =
        readTag ?? (path => Task.FromResult(LocalTagReader.Read(path)));

    /// <summary>
    /// Piso de <c>score</c> de MusicBrainz (0–100) para aceptar una grabación
    /// como fuente de álbum y año.
    ///
    /// <para>La búsqueda siempre devuelve el resultado de mayor puntaje aunque
    /// sea bajo, y usarlo sin piso hacía que dos canciones del mismo álbum real
    /// terminaran con álbumes distintos. Sin etiquetas locales es mejor dejar
    /// "Sin álbum" —que se puede revisar— que inventar uno. Un puntaje ausente
    /// cuenta como 0: se rechaza.</para>
    /// </summary>
    public const int MinimumMusicBrainzScore = 70;

    /// <summary>
    /// <paramref name="online"/> y <paramref name="lyrics"/> salen de las
    /// preferencias: sin conexión no se toca la red y solo se usa lo que trae el
    /// archivo más lo que se adivine del nombre.
    /// </summary>
    public async Task<TrackMetadata> EnrichAsync(
        LibraryItem item, bool online = true, bool lyrics = true,
        IReadOnlyList<CoverArtProvider>? coverArtOrder = null, bool deezerEnabled = true,
        CancellationToken ct = default)
    {
        TrackMetadata metadata = await _readTag(item.SourcePath).ConfigureAwait(false);

        FilenameGuesser.Guess guess = FilenameGuesser.For(item.SourcePath);
        metadata.Title ??= guess.Title;
        metadata.Artist ??= guess.Artist;

        if (!online) return metadata;

        MusicBrainzClient.Recording? recording;
        try
        {
            recording = await _musicBrainz
                .SearchRecordingAsync(metadata.Title, metadata.Artist, ct)
                .ConfigureAwait(false);
        }
        catch (Exception)
        {
            // Este camino corre al importar, en lote: un error de red no puede
            // detener la importación. Quien quiere saber del error usa
            // ReenrichAsync, que sí lo reporta.
            return metadata;
        }

        if (recording is null || (recording.Score ?? 0) < MinimumMusicBrainzScore) return metadata;

        ApplyRecording(metadata, recording);
        await ApplyReleaseAsync(metadata, recording, coverArtOrder, deezerEnabled, ct).ConfigureAwait(false);

        if (lyrics && metadata is { Title: { } title, Artist: { } artist })
        {
            try
            {
                metadata.SyncedLyrics = await _lrclib
                    .FetchSyncedLyricsAsync(title, artist, metadata.Album, ct: ct)
                    .ConfigureAwait(false);
            }
            catch (Exception) { /* ídem: sin letra, pero la canción entra */ }
        }

        return metadata;
    }

    /// <summary>
    /// "Buscar información" / "Buscar letra" del menú contextual (D-198). A
    /// diferencia de <see cref="EnrichAsync"/> —que parte de las etiquetas
    /// crudas del archivo— parte de la metadata <b>ya resuelta</b> del item,
    /// para no pisar una corrección que el usuario hizo en la pantalla de
    /// revisión. Igual que aquélla, solo llena huecos.
    ///
    /// <para>Y a diferencia de aquélla, <b>no se traga los errores de red</b>:
    /// devuelve un <see cref="EnrichmentOutcome"/> para que la pantalla pueda
    /// decir "no se encontró nada" o "falló la conexión", que no son lo
    /// mismo.</para>
    /// </summary>
    public async Task<(TrackMetadata Metadata, EnrichmentOutcome Outcome)> ReenrichAsync(
        LibraryItem item, TrackMetadata currentMetadata, bool fetchAlbumInfo, bool fetchLyrics,
        IReadOnlyList<CoverArtProvider>? coverArtOrder = null, bool deezerEnabled = true,
        CancellationToken ct = default)
    {
        TrackMetadata metadata = currentMetadata;
        var outcome = new EnrichmentOutcome();

        FilenameGuesser.Guess guess = FilenameGuesser.For(item.SourcePath);
        string? seedTitle = metadata.Title ?? guess.Title;
        string? seedArtist = metadata.Artist ?? guess.Artist;

        if (fetchAlbumInfo)
        {
            try
            {
                MusicBrainzClient.Recording? recording = await _musicBrainz
                    .SearchRecordingAsync(seedTitle, seedArtist, ct)
                    .ConfigureAwait(false);

                if (recording is not null && (recording.Score ?? 0) >= MinimumMusicBrainzScore)
                {
                    outcome.AlbumInfoFound = true;
                    ApplyRecording(metadata, recording);
                    await ApplyReleaseAsync(metadata, recording, coverArtOrder, deezerEnabled, ct)
                        .ConfigureAwait(false);
                }
            }
            catch (Exception ex)
            {
                outcome.NetworkErrorMessage = ex.Message;
            }
        }

        if (fetchLyrics && metadata is { Title: { } title, Artist: { } artist })
        {
            try
            {
                metadata.SyncedLyrics = await _lrclib
                    .FetchSyncedLyricsAsync(title, artist, metadata.Album, ct: ct)
                    .ConfigureAwait(false);
                outcome.LyricsFound = metadata.SyncedLyrics is not null;
            }
            catch (Exception ex)
            {
                // El primer error de red es el que se muestra; el segundo no
                // agrega nada útil para el usuario.
                outcome.NetworkErrorMessage ??= ex.Message;
            }
        }

        return (metadata, outcome);
    }

    private static void ApplyRecording(TrackMetadata metadata, MusicBrainzClient.Recording recording)
    {
        metadata.Title ??= recording.Title;
        metadata.Artist ??= recording.ArtistCredit?.FirstOrDefault()?.Name;
        metadata.MusicBrainzRecordingId ??= recording.Id;
    }

    private async Task ApplyReleaseAsync(
        TrackMetadata metadata, MusicBrainzClient.Recording recording,
        IReadOnlyList<CoverArtProvider>? coverArtOrder, bool deezerEnabled, CancellationToken ct)
    {
        if (recording.Releases?.FirstOrDefault() is not { } release) return;

        metadata.Album ??= release.Title;
        metadata.Year ??= release.Date is { Length: >= 4 } date ? date[..4] : null;
        metadata.MusicBrainzReleaseId ??= release.Id;

        if (metadata.CoverArtData is null)
            metadata.CoverArtData = await ResolveCoverArtAsync(
                release.Id, release.ReleaseGroup?.Id, metadata.Title, metadata.Artist,
                coverArtOrder ?? CoverArtProviderInfo.DefaultOrder, deezerEnabled, ct)
                .ConfigureAwait(false);
    }

    /// <summary>
    /// Prueba los proveedores en el orden que eligió el usuario y devuelve la
    /// primera imagen que aparezca. Cada cliente ya devuelve <c>null</c> solo
    /// cuando no corresponde intentarlo (sin clave, o sin resultado); acá solo
    /// hace falta saltear Deezer si el usuario lo apagó, y que un proveedor que
    /// falla no tumbe a los demás.
    /// </summary>
    private async Task<byte[]?> ResolveCoverArtAsync(
        string releaseId, string? releaseGroupId, string? title, string? artist,
        IReadOnlyList<CoverArtProvider> order, bool deezerEnabled, CancellationToken ct)
    {
        foreach (CoverArtProvider provider in order)
        {
            try
            {
                byte[]? data = provider switch
                {
                    CoverArtProvider.CoverArtArchive =>
                        await _coverArt.FetchFrontCoverAsync(releaseId, ct).ConfigureAwait(false),

                    CoverArtProvider.FanartTV when releaseGroupId is not null =>
                        await _fanartTV.FetchAlbumCoverAsync(releaseGroupId, ct).ConfigureAwait(false),

                    CoverArtProvider.Deezer when deezerEnabled && title is not null && artist is not null =>
                        await _deezer.FetchAlbumCoverAsync(title, artist, ct).ConfigureAwait(false),

                    _ => null
                };

                if (data is { Length: > 0 }) return data;
            }
            catch (Exception)
            {
                // Un proveedor caído no puede impedir que se pruebe el siguiente.
            }
        }

        return null;
    }
}
